src/cloud-sql-instance.ts

```diff
@@ -107,51 +115,118 @@ export class CloudSQLInstance {
+        if (this.activeConnection) {
+          if (currentRefreshId === this.scheduledRefreshID) {
+            this.scheduleRefresh();
+          }
+        } else {
+          throw err as Error;
+        }
+
+        // This refresh cycle has failed, releases ref to ongoingRefreshPromise
+        this.ongoingRefreshPromise = undefined;
+      });
+
+    // The rate limiter needs to be initialized _after_ assigning a ref
+    // to ongoingRefreshPromise in order to avoid race conditions with
+    // the forceRefresh check that ensures a refresh cycle is not ongoing
     await this.initializeRateLimiter();
+
+    await this.ongoingRefreshPromise;
   }
 
-  private async _refresh(): Promise<void> {
+  // The refreshValues method will perform all the necessary async steps
+  // in order to get a new set of values for an instance that can then be
+  // used to create new connections to a Cloud SQL instance. It throws in
+  // case any of the internal steps fails.
+  private async refreshValues(): Promise<RefreshableValues> {
     const rsaKeys: RSAKeys = await generateKeys();
     const metadata: InstanceMetadata =
       await this.sqlAdminFetcher.getInstanceMetadata(this.instanceInfo);
 
-    this.ephemeralCert = await this.sqlAdminFetcher.getEphemeralCertificate(
+    const ephemeralCert = await this.sqlAdminFetcher.getEphemeralCertificate(
       this.instanceInfo,
       rsaKeys.publicKey,
       this.authType
     );
-    this.host = selectIpAddress(metadata.ipAddresses, this.ipType);
-    this.privateKey = rsaKeys.privateKey;
-    this.serverCaCert = metadata.serverCaCert;
+    const host = selectIpAddress(metadata.ipAddresses, this.ipType);
+    const privateKey = rsaKeys.privateKey;
+    const serverCaCert = metadata.serverCaCert;
+
+    return {
+      ephemeralCert,
+      host,
+      privateKey,
+      serverCaCert,
+    };
+  }
+
+  private updateValues(nextValues: RefreshableValues): void {
+    const {ephemeralCert, host, privateKey, serverCaCert} = nextValues;
+
+    this.ephemeralCert = ephemeralCert;
+    this.host = host;
+    this.privateKey = privateKey;
+    this.serverCaCert = serverCaCert;
+  }
+
+  private scheduleRefresh(): void {
+    const refreshInterval = getRefreshInterval(
+      /* c8 ignore next */
+      String(this.ephemeralCert?.expirationTime)
+    );
 
-    this.scheduledRefreshID = setTimeout(() => {
-      this.refresh();
-    }, getRefreshInterval(this.ephemeralCert.expirationTime));
+    this.scheduledRefreshID = setTimeout(() => this.refresh(), refreshInterval);
   }
 
   cancelRefresh(): void {
     if (this.scheduledRefreshID) {
       clearTimeout(this.scheduledRefreshID);
     }
+    this.scheduledRefreshID = null;
+  }
+
+  // Mark this instance as having an active connection. This is important to
+  // ensure any possible errors thrown during a future refresh cycle should
+  // not be thrown to the final user.
+  setActiveConnection(): void {
+    this.activeConnection = true;
```

Comment:
What counts as an `activeConnection`? Does this ever get unset? Technically if the cert expires then would this no longer be active? I think we probably need a check of sorts like the `isValid` function in other languages. https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/blob/main/internal/cloudsql/instance.go#L110

Reply:
The `activeConnection` is set whenever a TLS Socket connection is stablished using data from a given `CloudSQLInstance`. I believe that with the new refresh logic that's all that I want to check in order to silence thrown errors but I could have a method named `isValid` just for the sake of connector alignment, that would make sense! What do you think?

Reply:
makes sense! I'll use `stablishedConnection instead